tuner/examples/punet/punet_autotune.py

```diff
@@ -0,0 +1,181 @@
+        target_dir = mlir_spec_path.resolve().parent.parent.parent
+        output_name = f"unet_candidate_{candidate_tracker.candidate_id}.vmfb"
+        command = [
+            "timeout",
+            "300s",
+            "./compile-punet-base.sh",
+            "./iree-compile",
+            "gfx942",
+            f"{mlir_spec_path.resolve()}",
+            "./punet.mlir",
+            "-o",
+            (target_dir / output_name).as_posix(),
+        ]
+        return command
+
+    def get_model_benchmark_command(
+        self, candidate_tracker: libtuner.CandidateTracker
+    ) -> list[str]:
+        unet_candidate_path = candidate_tracker.compiled_model_path
+        assert unet_candidate_path is not None
+
+        command = [
+            "timeout",
+            "180s",
+            "iree-benchmark-module",
+            f"--device={libtuner.DEVICE_ID_PLACEHOLDER}",
+            "--hip_use_streams=true",
+            "--hip_allow_inline_execution=true",
+            "--device_allocator=caching",
+            f"--module={unet_candidate_path.resolve()}",
+            "--parameters=model=punet.irpa",
+            "--function=main",
+            "--input=1x4x128x128xf16",
+            "--input=1xsi32",
+            "--input=2x64x2048xf16",
+            "--input=2x1280xf16",
+            "--input=2x6xf16",
+            "--input=1xf16",
+            "--benchmark_repetitions=5",
+        ]
+        return command
+
+
+def main():
+    args = libtuner.parse_arguments()
+    path_config = libtuner.PathConfig()
+    path_config.base_dir.mkdir(parents=True, exist_ok=True)
+    path_config.output_unilog.touch()
+    candidate_trackers: list[libtuner.CandidateTracker] = []
+    punet_client = PunetClient()
+    stop_after_phase: str = args.stop_after
+
+    print("Setup logging")
+    libtuner.setup_logging(args, path_config)
+    print(path_config.run_log, end="\n\n")
+
+    print("Validating devices")
+    libtuner.validate_devices(args.devices)
+    print("Validation successful!\n")
+
+    print("Generating candidates...")
+    candidates = libtuner.generate_candidates(args, path_config, candidate_trackers)
+    print(f"Stored candidates in {path_config.candidates_dir}\n")
+    if stop_after_phase == libtuner.ExecutionPhases.generate_candidates:
+        return
+
+    print("Compiling candidates...")
+    compiled_candidates = libtuner.compile_dispatches(
+        args, path_config, candidates, candidate_trackers, punet_client
+    )
+    print(f"Compiled files are stored in {path_config.compiled_dir}\n")
+    if stop_after_phase == libtuner.ExecutionPhases.compile_dispatches:
+        return
+
+    print("Benchmarking compiled candidates...")
+    top_candidates = libtuner.benchmark_dispatches(
+        args, path_config, compiled_candidates, candidate_trackers, punet_client
+    )
+    print(f"Stored results in {path_config.output_unilog}\n")
+    if stop_after_phase == libtuner.ExecutionPhases.benchmark_dispatches:
+        return
```

Comment:
This could be refactored into a MapReduce or other dataflow pattern. Good opportunity to scale out across local or distributed CPUs, GPUs, and machines. Fine to focus on a single local machine at first though.

Reply:
why we want to refactor it to the MapReduce pattern? We already have multiprocessing inside of each tuning phase, and separated phases makes more sense to me, because it's good to let all candidates finish and evaluate the overall state (i.e. compilation rate, selecting top winners, etc.), and then decide whether to move to benchmarking.

Reply:
You have a few steps here that are "embarrassingly parallel" (no dependencies between tasks, so they can run independently with no communication needed) and would likely benefit from running at a scale not possible on a single machine.

Multiprocessing still limits you to a single machine. MapReduce is one programming model that lets you scale out across multiple machines. You could, for example,
1. generate 100,000 candidates on a single machine
2. compile 1000 candidates each on 100 CPU machines
3. benchmark 1000 candidates each on 100 GPUs

During an optimization effort, it sounded like we were seeing different tuning results across different GPUs. We could either accept that tuning will be GPU-specific or try to instead find generally beneficial tunings across a whole fleet of devices. I would expect this sort of "search the problem space for promising solutions" tool to be possible to run both locally on a single machine and distributed.

Reply:
I'm not sure this has clear mapping to map-reduce. Specifically, the reduce step has to reconcile the differences between physical gpus. I'd say that we should consider it in the future, but this is outside of the scope of what the tuner can do today. @RattataKing could you open a tracking issue for this?

Reply:
Some inspiration:
* https://pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html
* https://keras.io/guides/keras_tuner/distributed_tuning/